wuvt/admin/auth/views.py

```diff
@@ -0,0 +1,242 @@
+@auth_manager.check_access('admin')
+def user_add():
+    if app.config['AUTH_METHOD'] != "local":
+        abort(404)
+
+    error_fields = []
+    if current_user.username != 'admin':
+        abort(403)
+
+    if request.method == 'POST':
+        username = request.form['username'].strip()
+
+        if len(username) <= 2:
+            error_fields.append('username')
+
+        if len(username) > 8:
+            error_fields.append('username')
+
+        if not username.isalnum():
+            error_fields.append('username')
+
+        if User.query.filter_by(username=username).count() > 0:
+            error_fields.append('username')
+
+        name = request.form['name'].strip()
+
+        if len(name) <= 0:
+            error_fields.append('name')
+
+        email = request.form['email'].strip()
+
+        if len(email) <= 3:
+            error_fields.append('email')
+
+        password = request.form['password'].strip()
+
+        if len(password) <= 0:
+            error_fields.append('password')
+
+        # Create user if no errors
+        if len(error_fields) <= 0:
+            db.session.add(User(username, name, email))
+            db.session.commit()
+            user = User.query.filter_by(username=username).first()
+            user.set_password(password)
+            db.session.commit()
+
+            flash("User added.")
+            return redirect(url_for('admin.users'), 303)
+
+    return render_template('admin/user_add.html', error_fields=error_fields)
+
+
+@bp.route('/users/<int:id>', methods=['GET', 'POST'])
+@auth_manager.check_access('admin')
+def user_edit(id):
+    if app.config['AUTH_METHOD'] != "local":
+        abort(404)
+
+    user = User.query.get_or_404(id)
+    error_fields = []
+
+    # Only admins can edit users other than themselves
+    if current_user.username != 'admin' and current_user.id != id:
+        abort(403)
+
+    if request.method == 'POST':
+        name = request.form['name'].strip()
+        if len(name) <= 0:
+            error_fields.append('name')
+
+        # You can't change a username or ID
+
+        pw = request.form['newpass'].strip()
+
+        email = request.form['email'].strip()
+        if len(email) <= 0:
+            error_fields.append('email')
+
+        # TODO allow users to be disabled
+
+        if len(error_fields) == 0:
+            # update user's: name, email
+            user.name = name
+            user.email = email
+
+            # if user entered a new pw
+            if len(pw) > 0:
+                user.set_password(pw)
+            # TODO reset password to pw
+
+            db.session.commit()
+
+            flash('User updated.')
+            return redirect(url_for('admin.users'), 303)
+
+    else:
+        return render_template('admin/user_edit.html', user=user,
+                               error_fields=error_fields)
+
+
+@bp.route('/users')
+@auth_manager.check_access('admin')
+def users():
+    if app.config['AUTH_METHOD'] != "local":
```

Comment:
Now that OIDC auth is a thing, it's probably worthwhile to display non-local users anyway